File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.helper(root, float("-inf"), float("inf"))
    
    def helper(self, tree, minValue, maxValue):
        if tree is None: return True
        if tree.val <= minValue or tree.val >= maxValue: return False
        isLeftValid = self.helper(tree.left, minValue, tree.val)
        isRightValid = self.helper(tree.right, tree.val, maxValue)
        return isLeftValid and isRightValid
        

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return helper(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    
    private boolean helper(TreeNode tree, long minValue, long maxValue) {
        if (tree == null) return true;
        if (tree.val <= minValue || tree.val >= maxValue) return false;
        boolean isLeftValid = helper(tree.left, minValue, tree.val);
        boolean isRightValid = helper(tree.right, tree.val, maxValue);
        return isLeftValid && isRightValid;
  }
}